### PR TITLE
[TR] [SLA] PIM 5078 SLA move categories

### PR DESCRIPTION
**Pull request by BitOne**
| Q                 | A
| ----------------- | ---
| Specs             | OK
| Behats            | OK
| Blue CI           | Not yet
| Changelog updated | Yes
| Review and 2 GTM  | Not yet

### CODE REVIEW

File: features/Context/Page/Base/Base.php [hunk 1]
```diff
@@ -377,4 +377,24 @@ public function hasSelectedLocale($locale)
...
+     * Works better than the standard dragTo
+     *
+     * @param NodeElement $element
+     * @param NodeElement $dropZone
+     */
+    public function dragElementTo(NodeElement $element, NodeElement $dropZone)
+    {
+        $session = $this->getSession()->getDriver()->getWebDriverSession();
+
+        $from = $session->element('xpath',$element->getXpath());
```
Comment (ClementGautier) [inline features/Context/Page/Base/Base.php]: space after comma

File: features/Context/Page/Base/Base.php [hunk 2]
```diff
@@ -377,4 +377,24 @@ public function hasSelectedLocale($locale)
...
+        $to = $session->element('xpath',$dropZone->getXpath());
```
Comment (ClementGautier) [inline features/Context/Page/Base/Base.php]: same

File: features/Context/Page/Category/CategoryView.php [hunk 3]
```diff
@@ -98,13 +98,39 @@ public function expandCategory($category)
...
     public function dragCategoryTo($category1, $category2)
     {
         $category1 = $this->findCategoryInTree($category1);
         $category2 = $this->findCategoryInTree($category2);
 
-        $category1->dragTo($category2);
+        $session = $this->getSession()->getDriver()->getWebDriverSession();
+
+        $from = $session->element('xpath',$category1->getXpath());
+        $to = $session->element('xpath',$category2->getXpath());
```
Comment (ClementGautier) [inline features/Context/Page/Category/CategoryView.php]: uncomment or remove this test

File: features/Context/Page/Base/Base.php [hunk 4]
```diff
@@ -377,4 +377,24 @@ public function hasSelectedLocale($locale)
...
+
+        $session->moveto(array('element' => $from->getID()));
+        $session->buttondown("");
```
Comment (sumbobyboys) [inline features/Context/Page/Base/Base.php]: please use single quotes

File: features/Context/Page/Category/CategoryView.php [hunk 5]
```diff
@@ -98,13 +98,39 @@ public function expandCategory($category)
...
         return $this;
+    }*/
+
+    /**
+     * @param string $category1
+     * @param string $category2
+     *
+     * @return CategoryView
+     */
+    public function dragCategoryTo($category1, $category2)
```
Comment (ClementGautier) [inline features/Context/Page/Category/CategoryView.php]: where is this function used ? (I'm not a big fan of the "return this")

File: features/Context/Page/Base/Base.php [hunk 6]
```diff
@@ -377,4 +377,24 @@ public function hasSelectedLocale($locale)
...
+        $session->moveto(array('element' => $to->getID()));
+        $session->buttonup("");
```
Comment (sumbobyboys) [inline features/Context/Page/Base/Base.php]: same

File: src/Akeneo/Bundle/ClassificationBundle/Doctrine/ORM/Repository/CategoryRepository.php [hunk 7]
```diff
@@ -372,4 +372,24 @@ public function getOrderedAndSortedByTreeCategories()
...
     }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function persistAsNextSiblingOf(CategoryInterface $node, CategoryInterface $prevSibling)
+    {
+        parent::persistAsNextSiblingOf($node, $prevSibling);
+
+        $this->_em->flush($node);
```
Comment (jjanvier) [inline src/Akeneo/Bundle/ClassificationBundle/Doctrine/ORM/Repository/CategoryRepository.php]: isn't the flush in the repo a little bit weird?

File: features/Context/Page/Base/Base.php [hunk 8]
```diff
@@ -377,4 +377,24 @@ public function hasSelectedLocale($locale)
 
         return true;
     }
+
+    /**
+     * Drag an element on another one.
```
Comment (rybus) [inline features/Context/Page/Base/Base.php]: Drags

File: features/Context/Page/Base/Base.php [hunk 9]
```diff
@@ -377,4 +377,24 @@ public function hasSelectedLocale($locale)
...
+     * @param NodeElement $dropZone
+     */
+    public function dragElementTo(NodeElement $element, NodeElement $dropZone)
+    {
+        $session = $this->getSession()->getDriver()->getWebDriverSession();
+
+        $from = $session->element('xpath',$element->getXpath());
+        $to = $session->element('xpath',$dropZone->getXpath());
+
+        $session->moveto(array('element' => $from->getID()));
```
Comment (phaseinducer) [inline features/Context/Page/Base/Base.php]: `[]` please

Comment (phaseinducer) [inline features/Context/Page/Base/Base.php]: I think it's `getId()`

File: features/Context/Page/Base/Base.php (hunk 4 above)
Comment (phaseinducer) [inline features/Context/Page/Base/Base.php]: simple quotes please

File: src/Akeneo/Bundle/ClassificationBundle/Doctrine/ORM/Repository/CategoryRepository.php (hunk 7 above)
Comment (ClementGautier) [inline src/Akeneo/Bundle/ClassificationBundle/Doctrine/ORM/Repository/CategoryRepository.php]: in term of SRP, I think this flush should be placed somewhere else no ?

File: features/category/drag_and_drop_a_category.feature [hunk 10]
```diff
@@ -8,11 +8,14 @@ Feature: Drag and drop a category
...
-  @javascript @info This scenario does not accurately describe dragging because of limited behat/selenium dragging capabilities
-  Scenario: Move category to a different position in the tree
+  Scenario: Move category to a different parent in the tree
     Given I am on the categories page
     And I select the "2014 collection" tree
     And I expand the "Summer collection" category
     And I expand the "Winter collection" category
     And I drag the "Summer collection" category to the "Winter boots" category
     Then I should see the "Summer collection" category under the "Winter collection" category
+    Given I refresh current page
```
Comment (phaseinducer) [inline features/category/drag_and_drop_a_category.feature]: "When" here imho

File: src/Akeneo/Bundle/ClassificationBundle/Doctrine/ORM/Repository/CategoryRepository.php (hunk 7 above)
Comment (BitOne) [inline src/Akeneo/Bundle/ClassificationBundle/Doctrine/ORM/Repository/CategoryRepository.php]: Good question, but I believe it's cleaner than in the controller. And we don't have category saver.

File: spec/Akeneo/Bundle/ClassificationBundle/Doctrine/ORM/Repository/CategoryRepositorySpec.php [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+use Akeneo\Component\Classification\Model\CategoryInterface;
+use Doctrine\Common\EventManager;
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\Driver\Statement;
+use Doctrine\ORM\AbstractQuery;
+use Doctrine\ORM\EntityManager;
+use Doctrine\ORM\Mapping\ClassMetadata;
+use Doctrine\ORM\QueryBuilder;
+use Doctrine\ORM\Query\Expr;
+use Gedmo\Tree\Configuration;
```
Comment (skeleton) [inline spec/Akeneo/Bundle/ClassificationBundle/Doctrine/ORM/Repository/CategoryRepositorySpec.php]: useless

File: src/Akeneo/Bundle/ClassificationBundle/Doctrine/ORM/Repository/CategoryRepository.php (hunk 7 above)
Comment (BitOne) [inline src/Akeneo/Bundle/ClassificationBundle/Doctrine/ORM/Repository/CategoryRepository.php]: And about SRP, I agree, but I didn't want to break to many things by introducing a CategorySaver in a SLA patch ;)

**Comment by jjanvier** [comment]
GTM as long as we rework the flush on the master

**Comment by ClementGautier** [comment]
same at @jjanvier when you will forward this patch to master please create a CategorySaver and use it

**Comment by ClementGautier** [comment]
GTM

**Pull request merged**